--- ../src-base/minecraft/net/minecraft/pathfinding/NodeProcessor.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/NodeProcessor.java
@@ -7,8 +7,8 @@
 
 public abstract class NodeProcessor
 {
-    protected IBlockAccess blockaccess;
-    protected EntityLiving entity;
+    protected volatile IBlockAccess blockaccess;
+    protected volatile EntityLiving entity;
     protected final IntHashMap<PathPoint> pointMap = new IntHashMap<PathPoint>();
     protected int entitySizeX;
     protected int entitySizeY;
@@ -29,8 +29,10 @@
 
     public void postProcess()
     {
-        this.blockaccess = null;
-        this.entity = null;
+        synchronized (this){
+            this.blockaccess = null;
+            this.entity = null;
+        }
     }
 
     protected PathPoint openPoint(int x, int y, int z)
