--- ../src-base/minecraft/net/minecraft/pathfinding/PathHeap.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/PathHeap.java
@@ -1,102 +1,89 @@
 package net.minecraft.pathfinding;
 
-public class PathHeap
-{
+public class PathHeap {
     private PathPoint[] pathPoints = new PathPoint[128];
-    private int count;
-
-    public PathPoint addPoint(PathPoint point)
-    {
-        if (point.index >= 0)
-        {
-            throw new IllegalStateException("OW KNOWS!");
-        }
-        else
-        {
-            if (this.count == this.pathPoints.length)
-            {
+    private volatile int count;
+    public PathPoint addPoint(PathPoint point) {
+        if (point.index.get() < 0) {
+            if (this.count == this.pathPoints.length) {
                 PathPoint[] apathpoint = new PathPoint[this.count << 1];
                 System.arraycopy(this.pathPoints, 0, apathpoint, 0, this.count);
                 this.pathPoints = apathpoint;
             }
 
             this.pathPoints[this.count] = point;
-            point.index = this.count;
+            point.index.set(this.count);
             this.sortBack(this.count++);
-            return point;
         }
+        return point;
     }
 
-    public void clearPath()
-    {
+    public void clearPath() {
         this.count = 0;
     }
 
-    public PathPoint dequeue()
-    {
+    public PathPoint dequeue() {
+        if (this.count<0){
+            return null;
+        }
         PathPoint pathpoint = this.pathPoints[0];
         this.pathPoints[0] = this.pathPoints[--this.count];
         this.pathPoints[this.count] = null;
 
-        if (this.count > 0)
-        {
+        if (this.count > 0) {
             this.sortForward(0);
         }
 
-        pathpoint.index = -1;
+        pathpoint.index.set(-1);
         return pathpoint;
     }
 
-    public void changeDistance(PathPoint point, float distance)
-    {
+    public void changeDistance(PathPoint point, float distance) {
         float f = point.distanceToTarget;
         point.distanceToTarget = distance;
 
-        if (distance < f)
-        {
-            this.sortBack(point.index);
+        if (distance < f) {
+            this.sortBack(point.index.get());
+        } else {
+            this.sortForward(point.index.get());
         }
-        else
-        {
-            this.sortForward(point.index);
-        }
     }
 
-    private void sortBack(int index)
-    {
+    private void sortBack(int index) {
+        if (index<0 || index>this.pathPoints.length){
+            return;
+        }
         PathPoint pathpoint = this.pathPoints[index];
         int i;
 
-        for (float f = pathpoint.distanceToTarget; index > 0; index = i)
-        {
+        for (float f = pathpoint.distanceToTarget; index > 0; index = i) {
             i = index - 1 >> 1;
             PathPoint pathpoint1 = this.pathPoints[i];
 
-            if (f >= pathpoint1.distanceToTarget)
-            {
+            if (f >= pathpoint1.distanceToTarget) {
                 break;
             }
 
             this.pathPoints[index] = pathpoint1;
-            pathpoint1.index = index;
+            pathpoint1.index.set(index);
         }
 
         this.pathPoints[index] = pathpoint;
-        pathpoint.index = index;
+        pathpoint.index.set(index);
     }
 
-    private void sortForward(int index)
-    {
+    private void sortForward(int index) {
+        if (index<0 || index>this.pathPoints.length){
+            return;
+        }
         PathPoint pathpoint = this.pathPoints[index];
         float f = pathpoint.distanceToTarget;
 
-        while (true)
-        {
+        while (true) {
             int i = 1 + (index << 1);
             int j = i + 1;
 
-            if (i >= this.count)
-            {
+            if (i >= this.count) {
                 break;
             }
 
@@ -105,47 +92,38 @@
             PathPoint pathpoint2;
             float f2;
 
-            if (j >= this.count)
-            {
+            if (j >= this.count) {
                 pathpoint2 = null;
                 f2 = Float.POSITIVE_INFINITY;
-            }
-            else
-            {
+            } else {
                 pathpoint2 = this.pathPoints[j];
                 f2 = pathpoint2.distanceToTarget;
             }
 
-            if (f1 < f2)
-            {
-                if (f1 >= f)
-                {
+            if (f1 < f2) {
+                if (f1 >= f) {
                     break;
                 }
 
                 this.pathPoints[index] = pathpoint1;
-                pathpoint1.index = index;
+                pathpoint1.index.set(index);
                 index = i;
-            }
-            else
-            {
-                if (f2 >= f)
-                {
+            } else {
+                if (f2 >= f) {
                     break;
                 }
 
                 this.pathPoints[index] = pathpoint2;
-                pathpoint2.index = index;
+                pathpoint2.index.set(index);
                 index = j;
             }
         }
 
         this.pathPoints[index] = pathpoint;
-        pathpoint.index = index;
+        pathpoint.index.set(index);
     }
 
-    public boolean isPathEmpty()
-    {
+    public boolean isPathEmpty() {
         return this.count == 0;
     }
 }
