--- ../src-base/minecraft/net/minecraft/world/gen/NoiseGeneratorImproved.java
+++ ../src-work/minecraft/net/minecraft/world/gen/NoiseGeneratorImproved.java
@@ -26,10 +26,7 @@
         this.yCoord = p_i45469_1_.nextDouble() * 256.0D;
         this.zCoord = p_i45469_1_.nextDouble() * 256.0D;
 
-        for (int i = 0; i < 256; this.permutations[i] = i++)
-        {
-            ;
-        }
+        for (int i = 0; i < 256; this.permutations[i] = i++) {}
 
         for (int l = 0; l < 256; ++l)
         {
