--- ../src-base/minecraft/net/minecraft/block/BlockBeacon.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBeacon.java
@@ -106,39 +106,35 @@
 
     public static void updateColorAsync(final World worldIn, final BlockPos glassPos)
     {
-        HttpUtil.DOWNLOADER_EXECUTOR.submit(new Runnable()
-        {
-            public void run()
+        HttpUtil.DOWNLOADER_EXECUTOR.submit(() -> {
+            Chunk chunk = worldIn.getChunkFromBlockCoords(glassPos);
+
+            for (int i = glassPos.getY() - 1; i >= 0; --i)
             {
-                Chunk chunk = worldIn.getChunkFromBlockCoords(glassPos);
+                final BlockPos blockpos = new BlockPos(glassPos.getX(), i, glassPos.getZ());
 
-                for (int i = glassPos.getY() - 1; i >= 0; --i)
+                if (!chunk.canSeeSky(blockpos))
                 {
-                    final BlockPos blockpos = new BlockPos(glassPos.getX(), i, glassPos.getZ());
+                    break;
+                }
 
-                    if (!chunk.canSeeSky(blockpos))
-                    {
-                        break;
-                    }
+                IBlockState iblockstate = worldIn.getBlockState(blockpos);
 
-                    IBlockState iblockstate = worldIn.getBlockState(blockpos);
-
-                    if (iblockstate.getBlock() == Blocks.BEACON)
+                if (iblockstate.getBlock() == Blocks.BEACON)
+                {
+                    ((WorldServer)worldIn).addScheduledTask(new Runnable()
                     {
-                        ((WorldServer)worldIn).addScheduledTask(new Runnable()
+                        public void run()
                         {
-                            public void run()
-                            {
-                                TileEntity tileentity = worldIn.getTileEntity(blockpos);
+                            TileEntity tileentity = worldIn.getTileEntity(blockpos);
 
-                                if (tileentity instanceof TileEntityBeacon)
-                                {
-                                    ((TileEntityBeacon)tileentity).updateBeacon();
-                                    worldIn.addBlockEvent(blockpos, Blocks.BEACON, 1, 0);
-                                }
+                            if (tileentity instanceof TileEntityBeacon)
+                            {
+                                ((TileEntityBeacon)tileentity).updateBeacon();
+                                worldIn.addBlockEvent(blockpos, Blocks.BEACON, 1, 0);
                             }
-                        });
-                    }
+                        }
+                    });
                 }
             }
         });
