--- ../src-base/minecraft/net/minecraft/util/CombatTracker.java
+++ ../src-work/minecraft/net/minecraft/util/CombatTracker.java
@@ -15,7 +15,7 @@
 
 public class CombatTracker
 {
-    private final List<CombatEntry> combatEntries = Lists.<CombatEntry>newArrayList();
+    private final List<CombatEntry> combatEntries = Lists.<CombatEntry>newCopyOnWriteArrayList();
     private final EntityLivingBase fighter;
     private int lastDamageTime;
     private int combatStartTime;
