--- ../src-base/minecraft/net/minecraft/client/entity/EntityOtherPlayerMP.java
+++ ../src-work/minecraft/net/minecraft/client/entity/EntityOtherPlayerMP.java
@@ -43,6 +43,7 @@
 
     public boolean attackEntityFrom(DamageSource source, float amount)
     {
+        net.minecraftforge.common.ForgeHooks.onPlayerAttack(this, source, amount);
         return true;
     }
 
@@ -122,9 +123,7 @@
 
         this.cameraYaw += (f1 - this.cameraYaw) * 0.4F;
         this.cameraPitch += (f - this.cameraPitch) * 0.8F;
-        this.world.profiler.startSection("push");
         this.collideWithNearbyEntities();
-        this.world.profiler.endSection();
     }
 
     public void sendMessage(ITextComponent component)
